### PR TITLE
socket refactor

### PR DESCRIPTION
**Pull request by tasinco**
Calculate the msg size only if we have active connections.
Send the length followed by the msg to avoid a byte copy.
Set the receiver to fetch a uint64, which is what was used for the send.

### CODE REVIEW

File: ipcs/socket/socket.go [hunk 1]
```diff
@@ -83,28 +83,39 @@ func (s *Socket) Listen() error {
...
+
 	var err error
 	errs := wrappers.Errs{}
+	// Prefix the message with an 8 byte length
+	lenBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(lenBytes, uint64(len(msg)))
+	// msg = append(lenBytes, msg...)
 	for _, conn := range conns {
-		if _, err = conn.Write(msg); err != nil {
+		if _, err = conn.Write(lenBytes); err != nil {
```
Comment (StephenButtolph) [inline ipcs/socket/socket.go:111]: Would this be cleaner as:
```
for _, byteSlice := range [][]byte{lenBytes, msg} {
    if _, err = conn.Write(msg); err != nil {
        s.removeConn(conn)
        errs.Add(fmt.Errorf("failed to write message to %s: %w", conn.RemoteAddr(), err))
        break
    }
}
```
?

File: ipcs/socket/socket.go [hunk 2]
```diff
@@ -83,28 +83,39 @@ func (s *Socket) Listen() error {
...
+	if len(conns) == 0 {
+		return nil
+	}
+
 	var err error
 	errs := wrappers.Errs{}
+	// Prefix the message with an 8 byte length
+	lenBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(lenBytes, uint64(len(msg)))
+	// msg = append(lenBytes, msg...)
```
Comment (StephenButtolph) [inline ipcs/socket/socket.go:109]: No need for this anymore.

File: ipcs/socket/socket.go [hunk 3]
```diff
@@ -83,28 +83,39 @@ func (s *Socket) Listen() error {
...
 
 	// Write to each connection
+	if len(conns) == 0 {
+		return nil
+	}
+
 	var err error
 	errs := wrappers.Errs{}
+	// Prefix the message with an 8 byte length
+	lenBytes := make([]byte, 8)
```
Comment (StephenButtolph) [inline ipcs/socket/socket.go:107]: I wonder if this could be allocated on the stack somehow.

Comment (tasinco) [reply ipcs/socket/socket.go:107]: [8]byte appears to break binary.BigEndian.PutUint64(lenBytes, uint64(len(msg)))

Comment (StephenButtolph) [reply ipcs/socket/socket.go:107]: If you change lenBytes to a `[8]byte` then you should be able to perform:
```
binary.BigEndian.PutUint64(lenBytes[:], uint64(len(msg)))
```
Similarly below you could range over:
```
[][]byte{lenBytes[:], msg}
```

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**